content/scala-cli.md

```diff
@@ -0,0 +1,150 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: implementation
+status: waiting-for-implementation
+title: SIP-NN Scala CLI as default Scala command
+---
+
+**By: Krzysztof Romanowski and  Scala CLI team**
+
+## History
+
+| Date          | Version            |
+|---------------|--------------------|
+| July 15th 2022 | Initial Draft      |
+
+## Summary
+
+We propose to replace current scripts that are installed as "Scala": `scala`, `scalac` and `scaladoc` with Scala CLI - a batteries included tool to interact with Scala. Scala CLI brings all the features that the commands above provide and expand them with incremental compilation, dependency management, packaging and much more. 
+
+
+## Motivation
+
+The current default Scala scripts: `scala`, `scalac`, `scaladoc` are quite limited. Beside that, the `scala` command that starts the Scala REPL is hardly used by non-power users. 
+
+The current scripts are lacking basic features such as support for resolving dependencies, incremental compilation or support for outputs other than JVM. This forces any user that wants to do anything more than just basic things to learn and use SBT, Mill or an other build tool and that adds to the complexity of learning Scala. 
+
+We observe that the current state of tooling in Scala is limiting creativity, with quite a high cost to create e.g. an application or a script with some dependencies that target Node.js. Many Scala developers are not choosing Scala for their personal projects, scripts, or small applications and we believe that the complexity of setting up a build tool is one of the reasons. 
```

Comment:
*NOTE*: Not part of the SIP committee, just a passerby with a comment.

> We observe that the current state of tooling in Scala is limiting creativity ... we believe that the complexity of setting up a build tool is one of the reasons

Firstly, I really enjoy `scala-cli`, and I believe it's a much needed tool. I'm really excited to see this here. However, I also think we should try to avoid hyperbole here. This is listed as one of the main points up here in motivation with the crux of the argument being what I quoted up above.

_Here is the minimal build tool setup in IDE including a dependency_

```
ThisBuild / scalaVersion := "3.1.3"

lazy val myproject = project
  .settings(
    libraryDependencies += "com.lihaoyi" %% "pprint" % "0.7.3"
  )
```
A single sbt command can also get you a fresh hello project that a user can just do `sbt run` and see `hello`. The same exists in Mill as well, which is arguably even easier.

_Here is the same with scala-cli_

```
//> using scala "3.1.1"
//> using lib "com.lihaoyi::pprint:0.7.3"
```
Is the difference above really what is limiting creativity in Scala? Is learning the basic sbt structure somehow more complex than learning using directives in scala-cli?

You do mention some good points here:

> with quite a high cost to create e.g. an application or a script with some dependencies that target Node.js

No doubt cross platform support is confusing in build tools, but seeing that this is a giant change in the ecoystem by changing the way someone first interacts with the language I think the SIP should include a couple things.

1. Why was this route chosen rather than improving some of these pain points in the existing tooling?
2. How is learning new concepts introduced by scala-cli such as learning directives and BSP easier to digest than other concepts in tools like sbt or Mill which you mention as a roadblock to creativity.

Mainly because investing time into a brand new tool is also an extremely high cost. Especially in a tooling ecosystem already starved of maintainers and contributors.

None of this means we shouldn't do it, but it just means we should be crystal clear about the _why_, and seeing that this is listed as one of the main motivations of this SIP I'd expect the questions up above also to be crystal clear in the document, and after reading this SIP they weren't answered for me.

Reply:
I can imagine that we in some future could have an official list of curated libs so that we could write something like:
```
//> using scala "3.1.1"
//> using pprint "0.7.3"
```
and *that* level of simplicity would be *really* nice from a beginner perspective, and this SIP is taking a big step towards that level of quality UX. (or is the new cool buzzword rather DX now :smile:  )

Reply:
> To avoid strange % and %% and : and :: that learners get intimidated by...

O for sure, but again, this doesn't address the main question I had about the effort going into a new tool, vs addressing these pain points in the existing tools. Take the example you have up above, and there actually are discussions about them both in [Mill](https://github.com/com-lihaoyi/mill/pull/1916) and in [sbt](https://github.com/sbt/sbt/discussions/6736). We shy away from talking about it, but in Scala we have a handful of tools that almost all have single maintainers, and sometimes barely that. A lot of pains points we feel _could_ be addressed given focus and people time. Imagine adding up all the hours of various devs that have gone into making `scala-cli`, and imagine putting all those hours towards exiting issues and problems. What would have a greater impact on the ecosystem?

Again, I want to reiterate that I'm not saying this necessarily as a reason _against_ `scala-cli`, I'm happily using `scala-cli`, but it's just a big glaring question mark still, is all the effort going into `scala-cli` (which also forces other tooling to put in effort to support this new tool) a net positive vs if all that effort was put into existing problems and tools. Maybe the answer is yes, starting from a clean slate offers us the greatest impact to make a change for the better in welcoming newcomers, addressing exiting problems, and pushing the boundaries of the existing tooling ecosystem. If that's the case, **then I 110% behind it**. Sign me up. For me personally, as someone who spends a lot of my time trying to help maintain other tools, answering the above question is most what I'm curious about.*

If this was just a project some person or company was working on because it was fun, or something they deemed worth it for them, cool. But I'm harping on the above so much because what we're proposing is a fundamental change to how someone is introduced to Scala, both from the command line and the editor. It's a big deal.

*<sub>I want to recognize that this isn't an easy question, or a new one. Balancing older stable tools with pain points and addressing them vs pushing the boundaries with new shiny ideas that don't fit within the confines of the existing tools. However I believe this just justifies putting the decision and answer to this question right at the forefront to show that it's really been discussed and thought about.</sub>

Reply:
> or is the new cool buzzword rather DX now 😄

😆 I'm so happy to see this being used here.

Reply:
I would like to emphasize this point with a concrete example of “build definition” for Scala CLI:

https://github.com/VirtusLab/typed-frames/blob/5b6ca450d9c14a0c668283f00e1847298a9af050/publish-conf.scala#L1-L8

I don’t think this is significantly better or simpler than the equivalent with sbt, gradle, mill, etc. It just looks the same: you supply the information to put in the pom.xml that will ultimately be published.

So, my suggestion would be to rephrase the sentence that raised this discussion, to position Scala CLI differently. I don’t think it improves much the landscape of build tools, but I think it improves significantly the current Scala interpreter (the one we get with the `scala` command line tool) because that one can’t be configured with directives.

Reply:
As far as I know, it lacks some features, see e.g. https://github.com/com-lihaoyi/Ammonite/issues/1093

Reply:
> @ckipp01 I can imagine that we in some future could have an official list of curated libs so that we could write something like:
> 
> ```
> //> using scala "3.1.1"
> //> using pprint "0.7.3"
> ```

Related: https://contributors.scala-lang.org/t/why-doesnt-scala-have-library-listing/5696

Reply:
May I humbly suggest to extend the section as
> Why decided to work on Scala CLI rather then improve existing tools like sbt, Mill _or Ammonite_?

For similar reasons as with sbt or Mill, Ammonite is an already existing tool that shares some similarities with `scala-cli`. It would be helpful to the wider Scala community to write down the reasons why it was better to start from scratch instead of building on top of Ammonite. Thank you :)

Reply:
> Is the difference above really what is limiting creativity in Scala? Is learning the basic sbt structure somehow more complex than learning using directives in scala-cli?

Since it is very much a first expression I believe it matters, yes. It's similar to arguing that `static public void main(args: String[])` is of so much of an issue, which is true, but since it's the first thing a newcomer sees it matters.